Add support for CREATE FUNCTION EXECUTE ON [MASTER | ALL SEGMENTS]

hlinnaka opened the pull request:
We already had a hack for the EXECUTE ON ALL SEGMENTS case, by setting
prodataaccess='s'. This exposes the functionality to users via DDL, and adds
the support for the EXECUTE ON MASTER case.

There was discussion on gpdb-dev about also supporting ON MASTER AND ALL
SEGMENTS, but that is not implemented yet. There is no handy "locus" in the
planner to represent that. There was also discussion about making a
gp_segment_id column implicitly available for functions, but that is also
not implemented yet.

There is no support for these new options in ORCA. Using any ON MASTER or
ON ALL SEGMENTS functions in a query cause ORCA to fall back. This is the
same as with the prodataaccess='s' hack that this replaces, but now that it
is more user-visible, it would be nice to teach ORCA about it.

The new options are only supported for set-returning functions. There is
mechanism to track where individual expressions should be executed, in the
planner, so that would require much more work. Likewise, they cannot be
used in target list. So you cannot do "SELECT on_master_func()", you must
do "SELECT * FROM on_master_funct()" instead. EXECUTE ON ANY functions,
which is the default, work the same as before.

hlinnaka commented:
Per discussion at https://groups.google.com/a/greenplum.org/d/msg/gpdb-dev/OuPc20lIdws/SunZLiVuAgAJ

hlinnaka commented:
I don't have the docs toolchain installed, and it's not covered by the pipeline, so if someone could verify that the docs build, that'd be great.

mkiyama commented:
Is this change for 6.0 or will it be backported to 5.x?

hlinnaka commented:
On 09/18/2017 06:36 PM, Mel Kiyama wrote:
> Is this change for 6.0 or will it be backported to 5.x?

For 6.0.

- Heikki

mkiyama commented:
UI still seems under discussion. Is that true?
https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/OuPc20lIdws

hlinnaka commented:
Pushed an update, to fix regression failures.

I didn't notice earlier, that the old behavior was that unless a function was marked as IMMUTABLE, the planner would always run it in the master. Restored that behavior. That explains why the functions in gp_toolkit were marked as IMMUTABLE, even though they clearly were not immutable. It was to allow the planner to execute them in the segments.

hlinnaka commented:
On 09/18/2017 07:43 PM, Mel Kiyama wrote:
> UI still seems under discussion. Is that true?
> https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/OuPc20lIdws

Yeah. The docs changes in this PR reflect what the code implements, but 
we may still decide to change the syntax, possibly as a separate PR.

> Also, the change is not reflected in the syntax diagram for the dita doc (xml)

Fixed, thanks!

- Heikki

jimmyyih commented on src/include/catalog/pg_proc.h:
```diff
@@ -4600,13 +4603,14 @@ DESCR("flag an extension's table contents to be emitted by pg_dump");
  * generated by running catullus.pl script. But since we cannot specify the
  * PRODATAACCESS column of pg_proc as SEGMENT ('s') using the CREATE FUNCTION
  * syntax, we need to create the DATA statement manually and place it here.
+ * FIXME comment
```
Finish your thoughts here in the "FIXME comment" or remove the line?

mkiyama commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_FUNCTION.xml:
```diff
@@ -46,7 +47,41 @@ for more information.</pd></plentry><plentry><pt>IMMUTABLE</pt><pt>STABLE</pt><p
+    <pd>
+      <codeph>EXECUTE ON ANY</codeph> (the default) indicates
+      that the function may be executed on any node, and it will return
+      the same result regardless of where it is executed.
```
I would replace "node" with "segment instance"

mkiyama commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_FUNCTION.xml:
```diff
@@ -46,7 +47,41 @@ for more information.</pd></plentry><plentry><pt>IMMUTABLE</pt><pt>STABLE</pt><p
+      <codeph>EXECUTE ON ALL SEGMENTS</codeph> indicates that the
+      function must be executed on all segments. The overall result
+      of the function is the UNION ALL of the results from all segments.
+    </pd>
```
For "segments" I would say "segment instances"
In the past, there was confusion between segment hosts and segment instances

mkiyama commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_FUNCTION.xml:
```diff
@@ -46,7 +47,41 @@ for more information.</pd></plentry><plentry><pt>IMMUTABLE</pt><pt>STABLE</pt><p
+      <codeph>ON MASTER</codeph> and <codeph>ALL SEGMENTS</codeph> are
+      only allowed on set-returning functions. Note that executing queries
+      that access any tables is only supported when a function is executed
+      in the master. Such functions should be marked as <codeph>EXECUTE
```
"is executed ~~in~~ on the master"

mkiyama commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_FUNCTION.xml:
```diff
@@ -46,7 +47,41 @@ for more information.</pd></plentry><plentry><pt>IMMUTABLE</pt><pt>STABLE</pt><p
+    <pt>EXECUTE ON ANY</pt>
+    <pt>EXECUTE ON MASTER</pt>
+    <pt>EXECUTE ON ALL SEGMENTS</pt>
+
```
Can  EXECUTE ON * be used with?
IMMUTABLE | STABLE | VOLATILE

If so, is there anything that to document about the interaction between the two?

Also, EXECUTE ON * are function attributes, correct?
If so are there any special limitations on where the attributes can be placed?

jimmyyih commented:
PR pipeline shows some ICW test failures relating to `gp_read_error_log` and removed/revised gp_toolkit functions.  In TINC, there are a couple references to `__gp_localid` and `__gp_masterid` in answer files that need to be removed.

hlinnaka commented:
On 09/18/2017 09:24 PM, Jimmy Yih wrote:
> jimmyyih commented on this pull request.
> 
> 
> 
>> @@ -4600,13 +4603,14 @@ DESCR("flag an extension's table contents to be emitted by pg_dump");
>    * generated by running catullus.pl script. But since we cannot specify the
>    * PRODATAACCESS column of pg_proc as SEGMENT ('s') using the CREATE FUNCTION
>    * syntax, we need to create the DATA statement manually and place it here.
> + * FIXME comment
> 
> Finish your thoughts here in the "FIXME comment" or remove the line?

Ah, thanks, I had forgotten about that loose end!

I added support for EXECUTE ON to catullus.pl, and moved the definition 
to pg_proc.sql. I added that as a separate commit, if you want to review 
just that part, but I'll squash it before pushing.

- Heikki

hlinnaka commented:
On 09/18/2017 10:16 PM, Mel Kiyama wrote:
> mkiyama commented on this pull request.
> 
> 
> 
>> @@ -46,7 +47,41 @@ for more information.</pd></plentry><plentry><pt>IMMUTABLE</pt><pt>STABLE</pt><p
>   <codeph>CREATE FUNCTION</codeph> for details. </pd></plentry><plentry><pt>[ EXTERNAL ] SECURITY INVOKER</pt><pt>[ EXTERNAL ] SECURITY DEFINER</pt><pd>Change whether the function is a security definer or not. The key
>   word <codeph>EXTERNAL</codeph> is ignored for SQL conformance. See <codeph>CREATE
>   FUNCTION</codeph> for more information about this capability.</pd></plentry>
> -        <plentry><pt>COST <varname>execution_cost</varname></pt>
> +
> +  <plentry>
> +    <pt>EXECUTE ON ANY</pt>
> +    <pt>EXECUTE ON MASTER</pt>
> +    <pt>EXECUTE ON ALL SEGMENTS</pt>
> +
> +    <pd>
> +      <codeph>EXECUTE ON ANY</codeph> (the default) indicates
> +      that the function may be executed on any node, and it will return
> +      the same result regardless of where it is executed.
> 
> I would replace "node" with "segment instance"

Hmm. In this case, it can also be executed in the master. Do we have a 
word for "master" or "any segment instance"? Or should I spell that out 
as "may be executed on the master, or any segment instance"?

- Heikki

hlinnaka commented:
Jimmy and Mel have already commented on the PR, and we're making good 
progress. But one problem came up, looking at the regression failures:

Currently, before the PR, you can do:

SELECT (gp_read_error_log('table')).errmsg;

In the PR, I was going to forbid using a an EXECUTE ON MASTER / ALL 
SEGMENTS function in the SELECT like that. It didn't seem like a big 
deal, because it is more natural to do "SELECT errmsg FROM 
gp_read_error_log('table')". The reason to forbid it is that things get 
a lot more complicated in the planner, if we tried to handle that. You 
can have function calls anywhere in the query. You might even have an 
expression that uses two functions, one defined as EXECUTE ON MASTER and 
another EXECUTE ON ALL SEGMENTS, so you would have to somehow split the 
execution of the single expression across nodes.

I was surprised to find that we actually have queries like the above in 
the regression tests. And we even use that syntax in the examples in the 
docs: 
https://gpdb.docs.pivotal.io/4350/admin_guide/load/topics/g-define-an-external-table-with-single-row-error-isolation.html. 
I was surprised, because I didn't think that the current 
"prodataaccess='s'" hack that we have used to flag functions that need 
to run on segments, would do anything when a function is called like 
that. And indeed it doesn't. To make such queries work anyway, we have 
this in the gp_read_error_log() function itself:

> 		/*
> 		 * Though this function is usually executed on segment, we dispatch
> 		 * the execution if it happens to be on QD, and combine the results
> 		 * into one set.
> 		 */
> 		if (Gp_role == GP_ROLE_DISPATCH)
> 		{ ... }

What to do? I think we have to keep supporting that, at least in the 
simple cases, for backwards-compatibility. And for a function like 
gp_truncate_error_log(), that just "performs" some action on all nodes, 
and you're not really interested in the result, "SELECT function()" is 
admittedly a more natural way to call it than "SELECT * FROM function()".

Maybe there's some way to support it that doesn't require contorting the 
planner too much. Perhaps we can handle the simple case, when there are 
no other relations involved, i.e. "SELECT function()" would work but 
"SELECT function() FROM table" would be rejected.

Thoughts? Any ideas?

- Heikki

mkiyama commented:
@hlinnaka 
Sorry, there is no term. for "master or any segment instance"
I think on "the master or any segment instance" is ok

hlinnaka commented:
Ok, I reworked the code to deal gracefully with functions in targetlist, like "SELECT func()". This hopefully fixes the regression failures.

hlinnaka commented:
On 09/18/2017 10:26 PM, Mel Kiyama wrote:
>> @@ -46,7 +47,41 @@ for more information.</pd></plentry><plentry><pt>IMMUTABLE</pt><pt>STABLE</pt><p
>   <codeph>CREATE FUNCTION</codeph> for details. </pd></plentry><plentry><pt>[ EXTERNAL ] SECURITY INVOKER</pt><pt>[ EXTERNAL ] SECURITY DEFINER</pt><pd>Change whether the function is a security definer or not. The key
>   word <codeph>EXTERNAL</codeph> is ignored for SQL conformance. See <codeph>CREATE
>   FUNCTION</codeph> for more information about this capability.</pd></plentry>
> -        <plentry><pt>COST <varname>execution_cost</varname></pt>
> +
> +  <plentry>
> +    <pt>EXECUTE ON ANY</pt>
> +    <pt>EXECUTE ON MASTER</pt>
> +    <pt>EXECUTE ON ALL SEGMENTS</pt>
> +
> 
> Can  EXECUTE ON * be used with?
> IMMUTABLE | STABLE | VOLATILE

Yes. I'm not sure sensible it is to mark a function as IMMUTABLE and 
also EXECUTE ON MASTER or EXECUTE ON ALL SEGMENTS. They are kind of 
contradictory, although I guess in principle you might have a function 
that only works in the master, e.g. because a library is missing from 
the other nodes, but which doesn't otherwise depend on where it's run.

> If so, is there anything that to document about the interaction between the two?

I'm not sure. One interaction between EXECUTE ON ANY and volatility is 
that if you use a function in the FROM clause, e.g. "SELECT ... FROM 
function()", the planner will always execute the function in the master 
unless it's also marked as IMMUTABLE. That's a pre-existing behavior and 
I kept that, since there might be applications that depend on that out 
there. That doesn't apply to functions used elsewhere in the query, only 
in the FROM clause, though. I'm not sure what, if anything, we should 
say about that in the documentation. It would be nice to document it, 
but OTOH, it's more of an implementation detail at the moment and I 
would not like to make it a hard promise to user, in case we want to 
change that in the future.

> Also, EXECUTE ON * are function attributes, correct?

Yes.

> If so are there any special limitations on where the attributes can be placed?

Nope, they work the same as STRICT, VOLATILE, etc.

Thanks for your comments! I changed the nomenclature on "node" and 
"segments", please have another look.

- Heikki

mkiyama commented on gpdb-doc/dita/ref_guide/sql_commands/CREATE_FUNCTION.xml:
```diff
@@ -121,7 +122,42 @@ to be executed with the privileges of the user that created it. The key
+
+    <pd>
+      <codeph>EXECUTE ON MASTER</codeph> indicates that the function
+      must be executed on the master host.
```
Minor tweak. I would say "master instance"
This is for the case where segment instances are on the same host as the master,
like a single node system.

mkiyama commented:
just a thought...
make IMMUTABLE | STABLE | VOLATILE attributes mutually exclusive to the EXECUTE * attributes?

Some guidance is needed as to when to use the attributes, like your comments above.

hlinnaka commented:
On 09/19/2017 08:57 PM, Mel Kiyama wrote:
> just a thought...
> make IMMUTABLE | STABLE | VOLATILE attributes mutually exclusive to the EXECUTE * attributes?

Hmm, no, I don't think that would be good. Even though there are some 
interactions, they're not really the same thing. For example, an 
IMMUTABLE function can be evaluated at planning time already, while a 
VOLATILE function needs to re-evaluated for every row in the query. 
That's orthogonal to *where* they're executed.

> Some guidance is needed as to when to use the attributes, like your comments above.

Yeah. Volatility and EXECUTE ON are mostly orthogonal. Roughly speaking, 
volatility means *when* the function needs to be evaluated, and EXECUTE 
ON means *where* it needs to be evaluated. Some combinations are quite 
far-fetched, it's hard to imagine a function that's IMMUTABLE but must 
be executed at a certain node, for example.

- Heikki

hlinnaka commented:
So, there's been a lot of churn here, but now all the regression tests are passing, and this PR is ready to be pushed as far as I know. Please do a final review, so that we can get this in.

Per Mel's comments, would be good to add more details to the documentation on when to use these options, but I think that can be done as a follow-up patch. I think the information that's included here is sufficient for a reference page.

hlinnaka commented:
On 09/20/2017 05:02 PM, Daniel Gustafsson wrote:
> danielgustafsson approved this pull request.
> 
> I can't see anything that I disagree with, nice work! 

Ok, pushed, thanks!

> I guess leaving psql tab completion out of this patch was intentional?

Uh, didn't even think of it, to be honest. Not sure, tab-completion is 
really nice. I wouldn't expect many users to use this option from the 
psql prompt, it's more of a tool for extension writers. We haven't added 
tab-completion for all the GPDB-specific syntax, so doesn't seem bad to 
leave this out as well.

- Heikki
